packages/vite/src/node/ssr/ssrRuntimeServer.ts

```diff
@@ -0,0 +1,357 @@
+  id: string,
+  patterns?: (string | RegExp)[] | true,
+) {
+  if (patterns == null) return false
+  if (patterns === true) return true
+  for (const ex of patterns) {
+    if (typeof ex === 'string') {
+      if (id.includes(`/node_modules/${ex}/`)) return true
+    } else {
+      if (ex.test(id)) return true
+    }
+  }
+  return false
+}
+
+function patchWindowsImportPath(path: string) {
+  if (path.match(/^\w:\\/)) return `file:///${slash(path)}`
+  else if (path.match(/^\w:\//)) return `file:///${path}`
+  else return path
+}
+
+let SOURCEMAPPING_URL = 'sourceMa'
+SOURCEMAPPING_URL += 'ppingURL'
+
+const VITE_NODE_SOURCEMAPPING_SOURCE = '//# sourceMappingConsumer=vite'
+const VITE_NODE_SOURCEMAPPING_URL = `${SOURCEMAPPING_URL}=data:application/json;charset=utf-8`
+// const VITE_NODE_SOURCEMAPPING_REGEXP = new RegExp(`//# ${VITE_NODE_SOURCEMAPPING_URL};base64,(.+)`)
+
+export function withInlineSourcemap(result: TransformResult): TransformResult {
+  const map = result.map
+  let code = result.code
+
+  if (!map || code.includes(VITE_NODE_SOURCEMAPPING_SOURCE)) return result
+
+  // to reduce the payload size, we only inline vite source map, because it's also the only one we use
+  const OTHER_SOURCE_MAP_REGEXP = new RegExp(
+    `//# ${SOURCEMAPPING_URL}=data:application/json[^,]+base64,(.+)`,
+    'g',
+  )
+  while (OTHER_SOURCE_MAP_REGEXP.test(code))
+    code = code.replace(OTHER_SOURCE_MAP_REGEXP, '')
+
+  const sourceMap = Buffer.from(JSON.stringify(map), 'utf-8').toString('base64')
+  result.code = `${code.trimEnd()}\n\n${VITE_NODE_SOURCEMAPPING_SOURCE}\n//# ${VITE_NODE_SOURCEMAPPING_URL};base64,${sourceMap}\n`
+
+  return result
+}
+
+export class ViteServer {
+  private fetchPromiseMap = new Map<string, Promise<ViteRuntimeFetchResult>>()
+  private transformPromiseMap = new Map<
+    string,
+    Promise<TransformResult | null | undefined>
+  >()
+
+  fetchCache = new Map<
+    string,
+    {
+      duration?: number
+      timestamp: number
+      result: ViteRuntimeFetchResult
+    }
+  >()
+
+  externalizeCache = new Map<string, Promise<string | false>>()
+
+  // debugger?: Debugger
+
+  constructor(
+    public server: ViteDevServer,
+    public options: ViteRuntimeServerOptions = {},
+  ) {
+    const ssrOptions = server.config.ssr
+    if (ssrOptions) {
+      options.deps ??= {}
+
+      // we don't externalize ssr, because it has different semantics in Vite
+      // if (ssrOptions.external) {
+      //   options.deps.external ??= []
+      //   options.deps.external.push(...ssrOptions.external)
+      // }
+
+      if (ssrOptions.noExternal === true) {
+        options.deps.inline ??= true
+      } else if (options.deps.inline !== true) {
+        options.deps.inline ??= []
+        const noExternal = Array.isArray(ssrOptions.noExternal)
+          ? ssrOptions.noExternal
+          : (ssrOptions.noExternal && [ssrOptions.noExternal]) || []
+        options.deps.inline.push(...noExternal)
+      }
```

Comment:
Do you see a way to merge `inline` and `noExternal`? What is `inline` providing differently? Maybe as part of this move, we should modify Vite's external/noExternal scheme so we are able to use the same across the board for SSR.

Reply:
I don't have enough knowledge of how Vite SSR works to answer that yet. For ViteNode `inline` means the code will be transformed with `transformRequest` and then it will run this code as if it was a source code.

Reply:
Ideally, I would like to have Vite SSR optimizer do the work, and `inline` will push values into `noExternal` and not vice versa.

Reply:
After the discussion in Discord I managed to simplify this part drastically. You can run the example with `node vite-node.mjs` in the root (I will remove example files when PR is ready to be merged)